compose: Migrate upload banner to new banner styling.

evykassirer opened the pull request:
This affects both the banner in the main compose box and the banner
in the message edit compose box. The use of ProgressBar has been
replaced with a more simple CSS (with light Javascript) solution.


Fixes #22524 (This is the final PR of three that fix this issue.)
Fixes #21480.

**Screenshots and screen captures:**

<img width="879" alt="image" src="https://user-images.githubusercontent.com/5634097/200466495-d7a1913f-8d37-419f-9e54-b11c6cf86fd5.png">

<img width="884" alt="image" src="https://user-images.githubusercontent.com/5634097/200466582-3c004cff-069c-42cf-8b06-363e4dce2867.png">

<img width="876" alt="image" src="https://user-images.githubusercontent.com/5634097/200467170-2f29b2bd-afe6-4f1c-be94-5fdddcd22113.png">

<img width="887" alt="image" src="https://user-images.githubusercontent.com/5634097/200963591-d00230b4-3128-44c3-bf75-b1d6f88e712a.png">

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

alya commented:
The banners in the screenshots look good to me (without comparing with the spec). @terpimost , any feedback from your side?

terpimost commented:
The right side of the loader shouldn't be with rounded corners:
<img width="214" alt="image" src="https://user-images.githubusercontent.com/1903309/204688076-1da5b5c4-9dc2-412a-92f0-0ac85fa1685b.png">
The hover and hit zone for the cancel button should be bigger than the icon itself:
<img width="78" alt="image" src="https://user-images.githubusercontent.com/1903309/204688155-856bdf1f-4919-4659-a9f6-ce8ad45bcfa3.png">
<img width="152" alt="image" src="https://user-images.githubusercontent.com/1903309/204688245-7da6bdb9-ae71-432b-9fbd-a4f63f2fce02.png">

It would be nice to see some banners where the text isn't all green/blue. 
<img width="892" alt="image" src="https://user-images.githubusercontent.com/1903309/204688799-bb186dd4-25c0-4f75-be11-ed69146c557e.png">

For example 
**Sent!** Scroll down to view your message. - the thing which isn't bold could be of a regular color.

Please check colors of the text and links. Looks like they are a bit different. Is there a opacity somewhere?
<img width="1074" alt="image" src="https://user-images.githubusercontent.com/1903309/204689248-42a030bf-bad8-4e22-b81e-1c47d056feb7.png">

evykassirer commented:
Iteration on this is blocked by #23421 and particularly #22754

evykassirer commented on frontend_tests/node_tests/upload.js:
```diff
@@ -159,47 +162,45 @@ test("get_item", () => {
 test("hide_upload_status", () => {
     $("#compose-send-button").prop("disabled", true);
-    $("#compose-send-status").addClass("alert-info").show();
+    $("#upload_banner").toggleClass("hidden", false);
```
probably will use .remove() and a render function instead of this

evykassirer commented on static/styles/dark_theme.css:
```diff
@@ -235,6 +235,50 @@
     }
 
+    div[id^="message-edit-upload-bar"],
+    #upload_banner {
```
TODO update dark mode styling

timabbott commented on static/templates/compose.hbs:
```diff
@@ -46,7 +46,13 @@
+                <p class="upload_msg"></p>
+                <a role="button" class="zulip-icon zulip-icon-close compose_banner_close_button"></a>
+            </div>
+        </div>
```
This looks like a singleton -- how does it work if multiple files are being uploaded at once?

evykassirer replied:
Ah yeah that will be fixed by [this](https://github.com/zulip/zulip/pull/23486#discussion_r1046448383). I'll mark this PR as no longer a draft when it's ready for review.

evykassirer replied:
It's actually currently implemented as a singleton on main. That's good that we can fix that here.

evykassirer replied:
Okay so making it not a singleton is a bit of a doozy and should probably happen in a separate PR. Do you have a preference for that happening first or this redesign happening first?

evykassirer replied:
Aha! So actually if there are multiple files being uploaded at once, they get combined into a single progress bar and the bar only goes away once the upload is complete. I'm okay keeping it like that for now, since it's a lot simpler than trying to find a unique identifier for the files and keep track of the banners

timabbott replied:
Yeah, I think a single progress bar for the bundle is probably fine in practice. There's probably a way to configure `uppy` to do something fancier if we need it, but that seems like a distraction from completing the migration to the new banner styling.

evykassirer commented on frontend_tests/node_tests/upload.js:
```diff
@@ -18,12 +18,11 @@ mock_esm("@uppy/core", {
-const compose_ui = mock_esm("../../static/js/compose_ui");
 mock_esm("../../static/js/csrf", {csrf_token: "csrf_token"});
 
+const compose_ui = zrequire("compose_ui");
```
this change can be a separate commit

evykassirer commented on frontend_tests/node_tests/upload.js:
```diff
@@ -207,52 +219,63 @@ test("upload_files", ({override, override_rewire}) => {
-        hide_upload_status_called = true;
-        assert.equal(config.mode, "compose");
-    });
+    // let hide_upload_status_called = false;
```
TODO: why isn't this being called properly?

evykassirer commented on static/templates/compose_banner/upload_banner.hbs:
```diff
@@ -0,0 +1,5 @@
+<div class="upload_banner compose_banner {{banner_type}}">
```
commit for moving the banner here and updating classnames

evykassirer commented on static/js/upload.js:
```diff
@@ -141,9 +166,9 @@ export function upload_files(uppy, config, files) {
-        }, 500);
+        // TODO: why was there a timeout here? the UI should change immediately on click,
+        // so is there a reason to delay anything else?
+        hide_upload_status(config);
```
make these two timeout removals also their own commit, with a clearer question involved

evykassirer commented on static/js/upload.js:
```diff
@@ -90,19 +97,36 @@ export function get_item(key, config) {
 }
 
-export function upload_files(uppy, config, files) {
+export async function upload_files(uppy, config, files) {
```
needed this async for tests

evykassirer commented on static/js/upload.js:
```diff
@@ -204,9 +229,13 @@ export function setup_upload(config) {
         },
     });
 
-    uppy.use(ProgressBar, {
```
a commit for removing the progress bar and moving to the homemade one

evykassirer commented on static/js/click_handlers.js:
```diff
@@ -337,13 +337,6 @@ export function initialize() {
             ui_util.blur_active_element();
         }
     });
-    $(".message_edit_form .send-status-close").on("click", function () {
```
This is no longer needed because the `send_status` container doesn't need to be hidden and the rest is handled in the `send_status_close_button` click handler

timabbott replied:
OK I notice that this has a `fadeOut` animation; I assume the other code path doesn't. I think we don't have a pattern of doing fadeOut when closing banners/popovers, so that's probably fine; after looking at the original UI, the `.stop(true)` part to send the progress bar to 0 before fading it out feels pretty weird anyway.

evykassirer commented:
remaining tasks:

- [ ] codecov issue
- [ ] fix upload bar to not be rounded while in progress (but still rounded near/at 100%)

evykassirer commented:
> The right side of the loader shouldn't be with rounded corners:

This has been fixed now

>  The hover and hit zone for the cancel button should be bigger than the icon itself

This is fixed in #23846

evykassirer commented:
@alya can you take a look for product review?

alya commented:
Hm, can we make the animation look continuous, not jumpy? This is testing on "Fast 3G":

![2023-02-01 23 47 35](https://user-images.githubusercontent.com/2090066/216263216-c095df16-b002-476e-b2e9-8cfdaebd4145.gif)

alya commented:
I also think there is a regression where on CZO I can start another upload while the first one is still going, and it doesn't seem like it works with this PR.

alya commented:
Things that worked fine for me in manual testing:
- I tried light and dark mode
- Adding another compose banner before or after the upload

alya commented:
I'm not sure about the colors for dark theme (they look a little greener than our other blues?), but maybe it'll look good with the full redesign. In any case, this part is definitely not a blocker for CZO testing.

evykassirer commented:
> I also think there is a regression where on CZO I can start another upload while the first one is still going, and it doesn't seem like it works with this PR.

Thanks for catching that! Found the issue and fixed it.

> can we make the animation look continuous, not jumpy? 

Ooo, love that idea. I added an animation -- I can't get it completely smooth because the upload tool only lets us know the progress every so often. Between any two progress updates there's now a 1s animation which I think looks a lot better.

![Kapture 2023-02-02 at 16 54 51](https://user-images.githubusercontent.com/5634097/216485382-dacd9e03-1268-4373-819a-d6fef16c05d7.gif)

> I'm not sure about the colors for dark theme (they look a little greener than our other blues?), but maybe it'll look good with the full redesign. In any case, this part is definitely not a blocker for CZO testing.

Seems good to wait until the background colors change, and then we could iterate on it with Vlad if needed!

evykassirer commented on frontend_tests/node_tests/upload.js:
```diff
@@ -29,7 +28,7 @@ const upload = zrequire("upload");
     run_test(label, (helpers) => {
         page_params.max_file_upload_size_mib = 25;
-        f(helpers);
+        return f(helpers);
```
needed for the `async` added below

evykassirer commented on frontend_tests/node_tests/upload.js:
```diff
@@ -492,7 +511,7 @@ test("uppy_events", ({override, override_rewire}) => {
     };
     upload.setup_upload({mode: "compose"});
-    assert.equal(Object.keys(callbacks).length, 5);
+    assert.equal(Object.keys(callbacks).length, 6);
```
The new one is `uppy.on("progress"` which we're using instead of having the progressbar handle it.

evykassirer commented:
Okay, now there's a weird transition around error banners, looking into it

evykassirer commented:
Alright good to go!

evykassirer commented on static/js/upload.js:
```diff
@@ -92,19 +97,47 @@ export function get_item(key, config) {
+    const $upload_banner = get_item("send_status", config);
+    if ($upload_banner.length) {
+        if (banner_type === "error") {
+            // Hide moving bar so that it doesn't do the 1s transition to 0
```
I don't super love this but couldn't think of a better thing to do.

Well, I guess I can take time in this PR to separate the error and upload banners out :p I'd rather do that in a followup though

timabbott replied:
This seems fine -- we also need to fix the handling of multiple simultaneous uploads, and that'll likely be a related refactor.

alya commented:
Looks good to me in manual testing -- thanks!

timabbott commented on tools/test-js-with-node:
```diff
@@ -204,6 +205,7 @@ EXEMPT_FILES = make_set(
         "static/js/unread.js",
         "static/js/unread_ops.js",
         "static/js/unread_ui.js",
+        "static/js/upload.js",
```
This commit seems to not have `git add`ed the new files?

evykassirer replied:
They're existing files. I just want to remove them from needing 100% codecov.

timabbott replied:
Oh, I somehow misread the commit message as "Extract" not "Exclude. OK.

As a nit, I'd have done just `node:` or "node tests" as the commit message title prefix -- the part before the colon is supposed to be the part of Zulip that's being modified, and "node tests" is that part -- "node codecov" is more in the weeds.

evykassirer replied:
Updated the commit message to say that.

timabbott commented on static/js/upload.js:
```diff
@@ -279,9 +277,7 @@ export function setup_upload(config) {
-            setTimeout(() => {
-                hide_upload_status(config);
-            }, 500);
+            hide_upload_status(config);
```
500ms seems excessive, but if you remove this timeout entirely, then users will never see the upload bar hit 100% -- I just tested with a 2MB file with no throttling on localhost, and with this PR, one sees the upload bar zoom along and then at like 60% it just disappears and the file is there. Some playing with it is likely required, but my guess is like a 50-300ms delay here will feel substantially nicer. I suspect if you watch closely with a major app like gmail, they have a similar delay before hiding.

evykassirer replied:
Ah! Cool, I can see doing a 100ms delay or something, and also adding a comment above the timeout for why it's there :)

evykassirer replied:
I'll keep the timeout from clicking the x button immediate.

evykassirer replied:
Updated!

evykassirer replied:
Actually ended up doing this:

```diff
@@ -141,9 +141,7 @@ export function upload_files(uppy, config, files) {
         compose_ui.autosize_textarea(get_item("textarea", config));
         uppy.cancelAll();
         get_item("textarea", config).trigger("focus");
-        setTimeout(() => {
-            hide_upload_status(config);
-        }, 500);
+        hide_upload_status(config);
     });
 
     for (const file of files) {
@@ -204,9 +202,14 @@ export function setup_upload(config) {
         },
     });
 
-    uppy.use(ProgressBar, {
-        target: get_item("send_status_identifier", config),
-        hideAfterFinish: false,
+    uppy.on("progress", (progress) => {
+        // When upload is complete, it resets to 0, which we don't want to do.
+        if (progress === 0) {
+            return;
+        }
+        $(`${get_item("send_status_identifier", config)} .moving_bar`).css({
+            width: `${progress}%`,
+        });
     });
 
     $("body").on("change", get_item("file_input_identifier", config), (event) => {
@@ -279,9 +282,11 @@ export function setup_upload(config) {
 
         const has_errors = get_item("send_status", config).hasClass("alert-error");
         if (!uploads_in_progress && !has_errors) {
+            // Hide upload status for 100ms after the 1s transition to 100%
+            // so that the user can see the progress bar at 100%.
             setTimeout(() => {
                 hide_upload_status(config);
-            }, 500);
+            }, 1100);
         }
     });
```

timabbott replied:
This seems to work pretty well playing with it in practice!

timabbott commented on static/js/upload.js:
```diff
@@ -307,6 +340,8 @@ export function setup_upload(config) {
             // The remaining errors are mostly frontend errors like file being too large
             // for upload.
+            // TODO: It would be nice to keep the other uploads going if one fails,
+            // and show both an error message and the upload bar.
```
Hmm, yeah, please open an issue for fixing this -- seems worth keeping track of.

evykassirer replied:
Opened #24287

timabbott commented on static/js/upload.js:
```diff
@@ -92,19 +97,47 @@ export function get_item(key, config) {
+            $moving_bar.hide();
+            $upload_banner.removeClass("info").addClass("error");
+            // Show it again once the animation is complete.
+            setTimeout(() => $moving_bar.show(), 1000);
```
Can you explain what this line is about? I worry about it doing weird things like re-showing the bar after the message is sent or the region has been dismissed, but maybe that's fine because an outer element will have been hidden?

evykassirer replied:
Hmm, it was to make sure it was visible for the future if the container becomes an upload (non-error) banner again, but I could also do one of these:

* render a new banner when transitioning from error to upload
* make sure the moving bar is visible when transitioning from error to upload

timabbott replied:
Hmm. Ultimately, we're going to want to render a new moving_bar for each file being uploaded, rather than being a singleton, right? So possibly that tweak would provide an opportunity to delete this.

evykassirer replied:
Yup that will let us delete it! The error bar being separate also simplifies it.

timabbott commented on static/js/upload.js:
```diff
@@ -19,7 +19,7 @@ export function feature_check($upload_button) {
 
 export function get_translated_status(file) {
-    const status = $t({defaultMessage: "Uploading {filename}…"}, {filename: file.name});
+    const status = $t({defaultMessage: "Uploading {filename}"}, {filename: file.name});
```
@alya I just wanted to check that removing the `...` is something we want to do -- that appears in the Figma design but I'm not sure it was explicitly discussed.

timabbott commented on static/styles/compose.css:
```diff
@@ -378,10 +378,53 @@
+        width: 0;
+        /* The progress updates seem to come every second or so,
+        so this is the smoothest it can probably get. */
+        transition: width 1s ease-in-out;
```
Looking at `node_modules/@uppy/progress-bar/src/style.scss`, it looks like they use a 0.4s animation duration in their own progress bar implementation; (and another for 0.2s on "height"; not sure what that does). How did you determine the 1s here? 

``` css
$ cat node_modules/@uppy/progress-bar/src/style.scss
@import '@uppy/core/src/_utils.scss';
@import '@uppy/core/src/_variables.scss';

.uppy-ProgressBar {
  /* no important */
  position: absolute;
  top: 0;
  left: 0;
  z-index: 10000;
  width: 100%;
  height: 3px;
  transition: height 0.2s;
}

.uppy-ProgressBar[aria-hidden=true] {
  /* no important */
  height: 0;
}

.uppy-ProgressBar-inner {
  width: 0;
  height: 100%;

  /* no important */
  background-color: $blue;
  box-shadow: 0 0 10px rgba($blue, 0.7);
  transition: width 0.4s ease;
}

.uppy-ProgressBar-percentage {
  position: absolute;
  top: 50%;
  left: 50%;

  /* no important */
  display: none;
  color: $white;
  text-align: center;
  transform: translate(-50%, -50%);
}
```

evykassirer replied:
```
 /* The progress updates seem to come every second or so,
        so this is the smoothest it can probably get. */
```

It's not perfect, but I wanted to get something close to a smooth moving bar, instead of jumping the bar every time a progress update comes in.

timabbott replied:
OK; it seems to feel OK when watching it go both with fast and slow uploads, so probably this is fine.

timabbott commented:
I did another round of testing and review, and I think the pending issues are:
* If we want this part of the app to feel nice, we probably do need to do the work to make these banners no longer a singleton. One very visible issue is that if you try uploading two files in a row, they fight over control of the progress bar and it looks weird.  Not a regression, but FYI for @alya.
* A decision on https://github.com/zulip/zulip/pull/23486#discussion_r1096282527 -- I don't have strong feelings there; but I want to make sure it's an explicit decision, since we'll be asking translators to do stuff. Also worth noting that probably if we make these upload banners no longer singletons as a follow-up, we're going to want each to say some version of "Uploading {filename}" so they don't all look identical.

timabbott commented:
OK I got a shoulder review from @alya confirming that the `...` removal was not an intentional design change -- just a wording detail not intended to be authoritative in the mockup. I dropped that commit, squashed the commit intended to be squashed, and marked this to merge once CI finishes. Thanks for making this happen @evykassirer!

I think there is probably a priority follow-up issue for making the upload widget do one banner per group of files uploaded, with "Uploading 5 files..." or "Uploading {filename}" as the strings, but I'll let @alya open that... this is a fairly visible element and it seems worth the investment to make it work properly while this banner implementation migration is fresh.

evykassirer commented:
>  One very visible issue is that if you try uploading two files in a row, they fight over control of the progress bar and it looks weird. 

I'm pretty sure there's just one jump backwards and then it updates on overall progress of uploading and not any individual file (there are two hooks, one for individual file progress and one for overall progress).
